modules/images/webp-uploads/load.php

```diff
@@ -4,41 +4,290 @@
+ *
+ * @return bool|int|WP_Error
  */
-function webp_uploads_filter_image_editor_output_format( $output_format, $filename, $mime_type ) {
-	// Only enable if the server supports WebP.
-	if ( ! wp_image_editor_supports( array( 'mime_type' => 'image/webp' ) ) ) {
-		return $output_format;
+function webp_uploads_generate_image_size( $attachment_id, $size, $mime ) {
+	$sizes    = wp_get_registered_image_subsizes();
+	$metadata = wp_get_attachment_metadata( $attachment_id );
+
+	if (
+		! isset( $metadata['sizes'][ $size ], $sizes[ $size ] )
+		|| ! is_array( $metadata['sizes'][ $size ] )
+		|| ! is_array( $sizes[ $size ] )
+	) {
+		return new WP_Error( 'image_mime_type_invalid_metadata', __( 'The image does not have a valid metadata.', 'performance-lab' ) );
 	}
 
-	// WebP lossless support is still limited on servers, so only apply to JPEGs.
-	if ( 'image/jpeg' !== $mime_type ) {
-		return $output_format;
+	// All subsizes are created out of the attached file.
+	$file = get_attached_file( $attachment_id );
+
+	// File does not exist.
+	if ( ! file_exists( $file ) ) {
+		return new WP_Error( 'image_file_size_not_found', __( 'The provided size does not have a valid image file.', 'performance-lab' ) );
 	}
 
-	// Skip conversion when creating the `-scaled` image (for large image uploads).
-	if ( preg_match( '/-scaled\..{3}.?$/', $filename ) ) {
-		return $output_format;
+	// Create the subsizes out of the attached file.
+	$editor = wp_get_image_editor( $file );
+
+	if ( is_wp_error( $editor ) ) {
+		return $editor;
+	}
+
+	$allowed_mimes = array_flip( wp_get_mime_types() );
+	if ( ! array_key_exists( $mime, $allowed_mimes ) || ! is_string( $allowed_mimes[ $mime ] ) ) {
+		return new WP_Error( 'image_mime_type_invalid', __( 'The provided mime type is not allowed.', 'performance-lab' ) );
+	}
+
+	if ( ! wp_image_editor_supports( array( 'mime_type' => $mime ) ) ) {
+		return new WP_Error( 'image_mime_type_not_supported', __( 'The provided mime type is not supported.', 'performance-lab' ) );
 	}
 
-	$output_format['image/jpeg'] = 'image/webp';
+	$extension = explode( '|', $allowed_mimes[ $mime ] );
+	$extension = reset( $extension );
 
-	return $output_format;
+	$width  = null;
+	$height = null;
+	$crop   = false;
+
+	if ( array_key_exists( 'width', $metadata['sizes'][ $size ] ) ) {
+		$width = $metadata['sizes'][ $size ]['width'];
+	} elseif ( array_key_exists( 'width', $sizes[ $size ] ) ) {
+		$width = $sizes[ $size ];
+	}
+
+	if ( array_key_exists( 'height', $metadata['sizes'][ $size ] ) ) {
+		$height = $metadata['sizes'][ $size ]['height'];
+	} elseif ( array_key_exists( 'width', $sizes[ $size ] ) ) {
+		$height = $sizes[ $size ];
+	}
+
+	if ( array_key_exists( 'crop', $sizes[ $size ] ) ) {
+		$crop = (bool) $sizes[ $size ]['crop'];
+	}
+
+	$editor->resize( $width, $height, $crop );
+	$filename = $editor->generate_filename( null, null, $extension );
+	$filename = preg_replace( '/-(scaled|rotated|imagifyresized)/', '', $filename );
```

Comment:
when the source file is large we should generate a WebP version of the `-scaled` image as well, we need that for markup. Currently only the jpeg version is generated.

Reply:
I think this is being discussed here: 

- https://github.com/WordPress/performance/issues/174

Reply:
Yeah, let's keep that out of this PR and review in #174 to follow up if needed.

Reply:
👍🏼